Mock API call

mdavis332 opened the pull request:
First pass at mocking Invoke-WebRequest so API calls are simulated. I've cobbled this together but haven't yet tested. If @cofonseca can test and confirm it's working, then it can be expanded to other Pester tests with Invoke-WebRequest commands.

mdavis332 committed:
Mock API call

First pass at mocking Invoke-WebRequest so API calls are simulated. I've cobbled this together but haven't yet tested. If @cofonseca can test and confirm it's working, then it can be expanded to other Pester tests with Invoke-WebRequest commands.

cofonseca commented:
Thanks for the contribution. Looks good at first glance. I'll test this out tomorrow when I have a bit more time and merge it in if everything looks good.

cofonseca commented:
The mock appears to be working, but the test fails. Here's a screenshot of the result I saw: 
http://prntscr.com/gt7cj7

cofonseca commented on Tests/New-HipchatUser.Tests.ps1:
```diff
@@ -1,16 +1,36 @@
+                Assert-MockCalled @MockParams
+            }
+            
+            It "Should return a 201 Status Code" {
```
This assertion seems to fail when I run it. http://prntscr.com/gt7cj7

mdavis332 committed:
Update New-HipchatUser.Tests.ps1 to resolve failures

Attempting to resolve failures described by @cofonseca indicated in screenshot https://prnt.sc/gt7cj7. Moved up the Assert-MockCalled function and added a test to ensure $result.StatusCode exists and is not null. Drew from https://powershell.org/forums/topic/mocking-with-pester/ and http://wahlnetwork.com/2016/11/28/building-powershell-unit-tests-pester-using-mock-commands/ as examples.

mdavis332 reviewed (commented):
Tried to address failure as described by @cofonseca

mdavis332 commented:
Adjusted the PR to try to address the failure indicated in your screenshot. I prob need to find time to (learn to) setup Pester in my own local env to better test.

cofonseca commented on Tests/New-HipchatUser.Tests.ps1:
```diff
@@ -1,16 +1,32 @@
 ﻿Import-Module PHipChatAdmin
```
PHipChatAdmin should be HipChatAdmin. The P must've been a typo that I made. Fixed this yesterday in a separate PR but if you could fix it here as well that would be awesome!

cofonseca commented on Tests/New-HipchatUser.Tests.ps1:
```diff
@@ -1,16 +1,32 @@
-        It "Should return a 201 Status Code" {
-            New-HipchatUser -FirstName 'Pester' -LastName 'Test' -ApiToken $ApiToken | 
-            Should Be '201'
+    InModuleScope PHipChatAdmin {
```
Another typo by me - this should just be HipChatAdmin, not PHipChatAdmin.

cofonseca commented on Tests/New-HipchatUser.Tests.ps1:
```diff
@@ -1,16 +1,32 @@
+            }
+            
+            It 'StatusCode field exists and not null' {
+                $result.StatusCode.ToString() | Should Not Be $null
```
Found the issue - this should be $Result.ToString() instead of $Result.StatusCode.ToString(). The reason for this is that the function already returns the StatusCode.

mdavis332 replied:
The Mock Invoke-WebRequest should return a hash table with keys id, Content, Status, and StatusCode. If $result is just a single string with a return code, I think it means my $global:result line is not using the mocked Invoke-WebRequest. Whatcha think?

cofonseca commented on Tests/New-HipchatUser.Tests.ps1:
```diff
@@ -1,16 +1,32 @@
+            }
+            
+            It "Should return a 201 Status Code" {
+                $result.StatusCode.ToString() | Should Be '201'
```
Same as above.

cofonseca commented:
@mdavis332 no worries! Think I found the issue. I commented on the PR. Let me know what you think.

mdavis332 committed:
Fix HipChatAdmin module name

Changed InModuleScope PHipChatAdmin to InModuleScope HipChatAdmin.

I'm still a bit confused. The $result should be a hash table with keys id, Content, Status, and StatusCode. Unless I'm missing something, $result shouldn't return StatusCode directly.

mdavis332 reviewed (commented):
Changed InModuleScope PHipChatAdmin to InModuleScope HipChatAdmin.

I'm still a bit confused. The $result should be a hash table with keys id, Content, Status, and StatusCode. Unless I'm missing something, $result shouldn't return StatusCode directly.

BTW, I can't tell if you get pinged on my PR comments, so I comment there and come back here and comment the same thing. Not sure if that's helping or adding to the noise :P

cofonseca commented:
@mdavis332 , sorry for the delay! I'm getting your pings, but I haven't had much free time, unfortunately.

I apologize if I'm incorrect, but this is what I think is happening. (Sorry in advance if this doesn't make sense... I'm not very good at explaining things).

In your mock, you're returning $returned, which is a hashtable, just like the real Invoke-WebRequest would inside of New-HipChatUser.ps1 (which is exactly what we want).

If we take a look at New-HipChatUser.ps1, we're returning a similar variable, called $Call, which includes the same hashtable that you're returning in $returned. $returned and $call both have a key called StatusCode, so $Returned.StatusCode and $Call.StatusCode are the same ('201').

In our test, we're running New-HipChatUser and storing the result in a variable, called $Result. At the end of New-HipChatUser.ps1, we're not returning $Call (the full hashtable), we're returning $Call.StatusCode (just one key), so the $Result should only be '201'.

At the end of the test, you're attempting to call $Result.StatusCode.ToString(), but .StatusCode doesn't exist, because New-HipChatUser.ps1 is already returning the .StatusCode, not the full hashtable. Changing your test to just look for $Result.ToString() will fix this.

When I tried doing that myself, I was able to get the test to pass. This is also confirmed because the API key in the test is fake, so there's no way the function would actually be making an API call. Your mock seems to be working perfectly!

Hope this all makes sense! Sorry for the short story.

mdavis332 committed:
Remove-Module & Re-import-Module before tests

Before test, remove any HipChatAdmin module in memory, then (re)Import-Module for the latest local HipChatAdmin. This ensures HipChatAdmin module in memory is what we expect.

mdavis332 reviewed (commented):
Ok, I think I figured out what was going on. Pester tests in my env seem to all complete successfully now. 

I was (and presumably you were) using the Pester test to test against the PSGallery in-memory HipChatAdmin module. Looking at the latest New-HipChatUser.ps1 in the repo, it was returning an object with lots of NoteProperties, including StatusCode. However, the PSGallery version was still just returning a single StatusCode value (no object). I added the remove-module and import-module at the top of the test to be sure it was using the latest code. 

GitHub seems to be complaining about a merge conflict that I'm not sure I understand or can resolve (perhaps you have to do that as the repo maintainer). Sorry for taking so long figuring this out. This is my biggest foray into making changes on GH so far and I'm learning, but slowly :) I'm still not comfortable with all the GH terminology and use, so trying to pick it up as I stumble along. 

Let me know if the latest changes now work for you.

cofonseca committed:
Merge branch 'master' into patch-3

cofonseca commented:
@mdavis332 Ah, I think you were right. My apologies! No worries on the delays - I'm still new to the process myself so I'm glad we're both learning.

I went ahead and resolved the conflict. This looks good. I'll merge it in. Thanks for your help!
